README.md
# ASGARDEX-CRYPTO

The ASGARDEX CRYPTO package is a crypto package used by all `ASGARDEX` clients.

ASGARDEX-CRYPTO encrypts a master phrase to a keystore. This keystore can then be exported to other ASGARDEX wallets or stored securely.

Users can export their phrase and import them into other wallets since it is a BIP39 compatible phrase.

## Design

Typically keystore files encrypt a `seed` to a file, however this is not appropriate or UX friendly, since the phrase cannot be recovered after the fact.

Crypto design:

[entropy] -> [phrase] -> [seed] -> [privateKey] -> [publicKey] -> [address]

Instead, ASGARDEX-CRYPTO stores the phrase in a keystore file, then decrypts and passes this phrase to other clients:

[keystore] -> ASGARDEX-CRYPTO -> [phrase] -> ChainClient

The ChainClients can then convert this into their respective key-pairs and addresses.
Users can also export their phrases after the fact, ensuring they have saved it securely. This could enhance UX onboarding since users aren't forced to write their phrases down immediately for empty or test wallets.

```js
// Crypto Constants for ASGARDEX
const AsgardexBIP39Phrase = 'asgardex'
const BIP44Path = "m/44'/931'/0'/0/0"
const cipher = 'aes-128-ctr'
const kdf = 'pbkdf2'
const prf = 'hmac-sha256'
const dklen = 32
const c = 262144
const hashFunction = 'sha256'
const meta = 'asgardex-keystore'
```

## Usage

### Basic usage

```js
import { generatePhrase, validatePhrase, encryptToKeyStore, decryptFromKeystore } from '../src/crypto'

const phrase = generatePhrase()
const isCorrect = validatePhrase(phrase)
const password = 'thorchain'
const keystore = await encryptToKeyStore(phrase, password)
const phraseDecrypted = await decryptFromKeystore(keystore, password)
```

Keystore Type

```js
export type Keystore = {
  address: string
  crypto: {
    cipher: string
    ciphertext: string
    cipherparams: {
      iv: string
    }
    kdf: string
    kdfparams: {
      prf: string
      dklen: number
      salt: string
      c: number
    }
    mac: string
  }
  id: string
  version: number
  meta: string
}
```

### Error handling

## Development

### Build

```bash
yarn build
```

### Tests

```bash
yarn test
```


CHANGELOG.md
# Changelog

## 0.1.0 (2020-06-13)

### Initial release


__tests__/crypto.test.ts
import { generatePhrase, validatePhrase, encryptToKeyStore, decryptFromKeystore } from '../src/crypto'

describe('Generate Phrase', () => {
  it('Generates 12-word phrase', () => {
    const phrase = generatePhrase()
    const words = phrase.split(' ')
    expect(words.length).toEqual(12)
  })
  it('Generates 24-word phrase', () => {
    const phrase = generatePhrase(24)
    const words = phrase.split(' ')
    expect(words.length).toEqual(24)
  })
})

describe('Validate Phrase', () => {
  it('Validates 12-word Phrase', () => {
    const phrase = generatePhrase()
    const correctPhrase = validatePhrase(phrase)
    expect(correctPhrase).toBeTruthy()
  })
  it('Validates 24-word Phrase', () => {
    const phrase = generatePhrase(24)
    const correctPhrase = validatePhrase(phrase)
    expect(correctPhrase).toBeTruthy()
  })
  it('Invalidates phrase', () => {
    const phrase = 'flush viable fury sword mention dignity ethics secret nasty gallery teach wrong'
    const incorrectPhrase = validatePhrase(phrase)
    expect(incorrectPhrase).toBeFalsy()
  })
})

describe('Export Keystore', () => {
  it('Generates Correct Keystore', async () => {
    const phrase = 'flush viable fury sword mention dignity ethics secret nasty gallery teach fever'
    const password = 'thorchain'
    const keystore = await encryptToKeyStore(phrase, password)
    expect(keystore.address).toEqual('thor1cge9rv0s3vyepx3gpa9f58j4e8ltxm580lnrpv')
    expect(keystore.crypto.cipher).toEqual('aes-128-ctr')
    expect(keystore.crypto.kdf).toEqual('pbkdf2')
    expect(keystore.crypto.kdfparams.prf).toEqual('hmac-sha256')
    expect(keystore.crypto.kdfparams.c).toEqual(262144)
    expect(keystore.version).toEqual(1)
    expect(keystore.meta).toEqual('asgardex-keystore')
  })
})

describe('Import Keystore', () => {
  it('Decrypts Keystore', async () => {
    const phrase = generatePhrase()
    const password = 'thorchain'
    const keystore = await encryptToKeyStore(phrase, password)
    const phraseDecrypted = await decryptFromKeystore(keystore, password)
    expect(phraseDecrypted).toEqual(phrase)
  })
})


lib/src/utils.d.ts
/// <reference types="node" />
export declare const getBytes: (string: string) => number[];
export declare const ab2hexstring: (arr: number[]) => string;
export declare const sha256ripemd160: (hex: string) => string;
export declare const encodeAddress: (value: string | Buffer, prefix?: string, type?: BufferEncoding) => string;
export declare const createAddress: (publicKey: number[]) => string;
export declare const pbkdf2Async: (passphrase: string | Buffer | Uint8Array | Uint8ClampedArray | Uint16Array | Uint32Array | Int8Array | Int16Array | Int32Array | Float32Array | Float64Array | DataView, salt: string | Buffer | Uint8Array | Uint8ClampedArray | Uint16Array | Uint32Array | Int8Array | Int16Array | Int32Array | Float32Array | Float64Array | DataView, iterations: number, keylen: number, digest: string) => Promise<Buffer>;


src/crypto.ts
const bip39 = require('bip39')
const crypto = require('crypto')
import { blake256 } from 'foundry-primitives'
const { v4: uuidv4 } = require('uuid')
import { pbkdf2Async, createAddress } from './utils'
const HDKey = require('hdkey')

// Constants
const AsgardexBIP39Phrase = 'asgardex'
const BIP44Path = "m/44'/931'/0'/0/0"
const cipher = 'aes-128-ctr'
const kdf = 'pbkdf2'
const prf = 'hmac-sha256'
const dklen = 32
const c = 262144
const hashFunction = 'sha256'
const meta = 'asgardex-keystore'

// Interfaces

export type Keystore = {
  address: string
  crypto: {
    cipher: string
    ciphertext: string
    cipherparams: {
      iv: string
    }
    kdf: string
    kdfparams: {
      prf: string
      dklen: number
      salt: string
      c: number
    }
    mac: string
  }
  id: string
  version: number
  meta: string
}

export const generatePhrase = (size = 12): string => {
  const entropy = size == 12 ? 128 : 256
  const phrase = bip39.generateMnemonic(entropy)
  return phrase
}

export const validatePhrase = (phrase: string): boolean => {
  return bip39.validateMnemonic(phrase)
}

export const getSeed = (phrase: string): string => {
  const words = phrase.split(' ')
  if (words.length != 12 && words.length != 24) {
    throw new Error('invalid phrase')
  }
  const seed = bip39.mnemonicToSeedSync(phrase, AsgardexBIP39Phrase)
  return seed
}

export const getAddress = (phrase: string): string => {
  const seed = getSeed(phrase)
  const hdkey = HDKey.fromMasterSeed(Buffer.from(seed, 'hex'))
  const childkey = hdkey.derive(BIP44Path)
  const address = createAddress(childkey._publicKey)
  return address
}

export const encryptToKeyStore = async (phrase: string, password: string): Promise<Keystore> => {
  const ID = uuidv4()
  const addr = getAddress(phrase)
  const salt = crypto.randomBytes(32)
  const iv = crypto.randomBytes(16)
  const kdfParams = {
    prf: prf,
    dklen: dklen,
    salt: salt.toString('hex'),
    c: c
  }
  const cipherParams = {
    iv: iv.toString('hex')
  }

  const derivedKey = await pbkdf2Async(Buffer.from(password), salt, kdfParams.c, kdfParams.dklen, hashFunction)
  const cipherIV = crypto.createCipheriv(cipher, derivedKey.slice(0, 16), iv)
  const cipherText: any = Buffer.concat([cipherIV.update(Buffer.from(phrase, 'utf8')), cipherIV.final()])
  const mac = blake256(Buffer.concat([derivedKey.slice(16, 32), Buffer.from(cipherText, 'hex')]))

  const cryptoStruct = {
    cipher: cipher,
    ciphertext: cipherText.toString('hex'),
    cipherparams: cipherParams,
    kdf: kdf,
    kdfparams: kdfParams,
    mac: mac
  }

  const keystore = {
    address: addr,
    crypto: cryptoStruct,
    id: ID,
    version: 1,
    meta: meta
  }

  return keystore
}

export const decryptFromKeystore = async (keystore: Keystore, password: string): Promise<string> => {
  const kdfparams = keystore.crypto.kdfparams
  try {
    const derivedKey = await pbkdf2Async(
      Buffer.from(password),
      Buffer.from(kdfparams.salt, 'hex'),
      kdfparams.c,
      kdfparams.dklen,
      hashFunction
    )

    const ciphertext = Buffer.from(keystore.crypto.ciphertext, 'hex')
    const mac = blake256(Buffer.concat([derivedKey.slice(16, 32), ciphertext]))

    if (mac !== keystore.crypto.mac) {
      return Promise.reject('invalid password')
    }
    const decipher = crypto.createDecipheriv(
      keystore.crypto.cipher,
      derivedKey.slice(0, 16),
      Buffer.from(keystore.crypto.cipherparams.iv, 'hex')
    )

    const phrase = Buffer.concat([decipher.update(ciphertext), decipher.final()])
    return Promise.resolve(phrase.toString('utf8'))
  } catch (error) {
    return Promise.reject(error)
  }
}
